### PR TITLE
Standardize error classes and automatically add span tags

### PR DESCRIPTION
**Pull request by danielamiao**

### CODE REVIEW

**Comment by bhs** [comment]
(discussed offline... waiting for a unittest of the function that has the interesting logic in it)

File: go/otgrpc/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/stretchr/testify/assert"
+
+	"github.com/opentracing/opentracing-go/mocktracer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	firstCode = codes.OK
+	lastCode  = codes.DataLoss
```
Comment (danielamiao) [inline go/otgrpc/errors_test.go:15]: I couldn't find a better way to iterate through all the GRPC error codes, there doesn't seem to be a native way to do it.

**Comment by bhs** [comment]
@danielamiao LGTM, but there are conflicts to resolve. :-/

**Comment by danielamiao** [comment]
All resolved! :)

**Comment by danielamiao** [comment]
I signed it!

File: go/otgrpc/client.go [hunk 2]
```diff
@@ -72,8 +72,8 @@ func OpenTracingClientInterceptor(tracer opentracing.Tracer, optFuncs ...Option)
 				clientSpan.LogFields(log.Object("gRPC response", resp))
 			}
 		} else {
-			clientSpan.LogFields(log.String("event", "error"), log.String("message", err.Error()))
-			ext.Error.Set(clientSpan, true)
+			SetSpanTags(clientSpan, err, true)
+      clientSpan.LogFields(log.String("event", "error"), log.String("message", err.Error()))
```
Comment (annismckenzie) [inline go/otgrpc/client.go:76]: Forgot to run `gofmt` on save? This indentation looks weird.

Comment (bhs) [reply go/otgrpc/client.go:76]: @annismckenzie thanks for catching that. I did a `go fmt ./...` and pushed master directly.

Comment (danielamiao) [reply go/otgrpc/client.go:76]: oh yeah sorry about that, I think it was me resolving conflicts via the Github web UI and it messed up the indentation, thanks for catching and fixing @annismckenzie and @bhs !

File: go/otgrpc/errors.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	code := codes.Unknown
+	if s, ok := status.FromError(err); ok {
+		code = s.Code()
+	}
+	span.SetTag("response_code", code)
+	span.SetTag("response_class", c)
+	if err == nil {
+		return
+	}
+	if client || c == ServerError {
```
Comment (aybabtme) [inline go/otgrpc/errors.go:66]: This seems to have significantly changed how and what errors are reported on spans, but no rationale is given on the PR as to why this is done.  Maybe it's obvious to others, but is it possible to get an explanations why this switch is the desired behavior?  I feel that `ext.Error.Set(span, true)` should be called as long as `err != nil`, no matter if the span was said to be a client span or detected to be a server span.

Comment (iheanyi) [reply go/otgrpc/errors.go:66]: Additionally, there's an assumption that all the errors here will be a gRPC error. So let's say there's an error at the database level in the server and its returned, it won't be of type `ServerError` and therefore won't be reported when it should be. I don't think we want database errors going undetected in that case.

**Pull request merged**